### ISSUE TITLE
Discrepancy between narrowed type and revealed type

### ISSUE THREAD

**Issue by Gobot1234** [microsoft/pyright#5944]
**Describe the bug**
Pyright and pylance disagree about the type of a symbol in one of my files of which to me at least looks somewhat like (**thought this code type checks fine**) 
```py
from typing import Any, NewType

AssetID = NewType("AssetID", int)


class HasIntID:
    id: int


class HasAssetIDID:
    id: AssetID


x: HasIntID | HasAssetIDID = Any

if isinstance(x, HasAssetIDID):
    reveal_type(x.id)  # reveals int 🤔
```
I can't seem to create a minimal repro here so unfortunately you're going to need to clone https://github.com/Gobot1234/steam.py and install the dependancies using `poetry install` 
Then going to https://github.com/Gobot1234/steam.py/blob/336064153a1fea5c751a5c5f60a71c1afe4b0c7f/steam/ext/csgo/state.py#L197 you should see a discrepancy between the 2 types, causing a type error as shown blow 
<img width="890" alt="image" src="https://github.com/microsoft/pyright/assets/50501825/d62281da-75fe-4277-90f1-68fb8e48bcbd">

**VS Code extension or command-line**
Pylance v2023.9.10

**Comment by erictraut** [comment]
Thanks for the bug report. 

This is an interesting hole in pyright's code flow engine, the logic used for computing narrowed types of local expressions. In this case, the code flow engine is asked to determine the narrowed type of `gc_item.id`. This type depends on the type of `gc_item`, so if `gc_item` is modified at some point in the code flow, any previously-narrowed type of `gc_item.id` must be discarded. The code flow engine looks for any assignments to `gc_item` to determine whether it was modified. In this example, however, there were no assignments. It was still "modified" in that it was narrowed (using an `isinstance` type guard). I'll need add code to detect this and discard the previously-narrowed type of `gc_item.id` in this case.

**Comment by Gobot1234** [comment]
Oh whilst this is still here as well, is https://github.com/Gobot1234/steam.py/blob/8e69746bd718e2ea3ae63ae47b81340c7a9ee1b4/steam/ext/commands/cog.py#L243 also another similar thing, Group is a subclass of Command so this should be possible to my knowledge however, pyright reports command in this branch as Never

**Comment by erictraut** [comment]
Looking closer into this, I think the problem might be a problem with your code. 

```python
    async def update_backpack(self, *gc_items: base.Item, is_cache_subscribe: bool = False) -> Backpack:
        ...

        gc_item: base.Item | CasketItem
        for gc_item in gc_items:
            if not isinstance(gc_item, CasketItem) and gc_item.id in self.casket_items:
                reveal_type(gc_item.id)

            elif isinstance(gc_item, CasketItem):
                reveal_type(gc_item.id)
                self.waiting_for_casket_items[gc_item.id]

        return backpack
```

The `update_backpack` method declares the type of the local variable `gc_item` to be `base.Item | CasketItem`. It is assigned value of type `base.Item` in the `for` statement. The second `isinstance` type guard narrows the type of `gc_item` to a synthesized intersection between `base.Item` and `CasketItem`. This is because these two types have no overlap, so the only way this `isinstance` conditional will succeed is if `gc_item` derives from a class that subclasses from both `base.Item` and `CasketItem`. However, these two classes both declare an attribute named `id`, and the type doesn't match between the two classes. That means any class that derives from both would have a type violation. The type of the expression `gc_item.id` depends on the on the MRO, and that depends (somewhat arbitrarily) on whether the synthesized intersection type is `<subclass of base.Item and CasketItem>` or `<subclass of CasketItem and base.Item>`.

Please review this code and let me know if it's doing what you expect. If so, I'm curious to understand the intent. It looks buggy to me.

**Comment by Gobot1234** [comment]
I'd say the code is more like 
```py
    async def update_backpack(self, *gc_items: base.Item, is_cache_subscribe: bool = False) -> Backpack:
        ...

        gc_item: base.Item | CasketItem
        for gc_item in gc_items:
            if bool():
                gc_item = CasketItem()  # strictly this is Any coming from update_class
                # however just CasketItem() is more like the actual behaviour here

            if not isinstance(gc_item, CasketItem) and gc_item.id in self.casket_items:
                reveal_type(gc_item.id)

            elif isinstance(gc_item, CasketItem):
                reveal_type(gc_item.id)
                self.waiting_for_casket_items[gc_item.id]

        return backpack
```
So intersections shouldn't be strictly necessary or it would be inconsistent with other pyright behaviour. e.g.
```py
def foo(x: int | str):
    if isinstance(x, str):
        reveal_type(x)  # str not str & int
```
even still the discrepancy between pylance and pyright on the revealed type isn't explained

**Comment by erictraut** [comment]
Looking at the latest steam.py code, I'm not able to repro the issue reported at the top of this thread. If you're still able to repro it, please open a new issue.

The second issue that you reported in this thread (in `cog.py`) was a legitimate bug in the `isinstance` type narrowing logic. This will be fixed in the next release.

**Comment by Gobot1234** [comment]
Thanks for fixing both!

**Comment by erictraut** [comment]
This is included in pyright 1.1.332, which I just published. It will also be included in a future release of pylance.
